tsl/src/compression/api.c

```diff
@@ -1075,128 +1061,692 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+		/* fetch number of rows present in current segment of compressed tuple */
+		AttrNumber meta_count =
+			get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_COUNT_NAME);
+		*metacount_val = values[meta_count - 1];
+	}
+	pfree(values);
+	pfree(nulls);
+}
+
+static int
+decompress_in_tuplestore(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+						 Tuplesortstate *merge_tuplestore)
+{
+	Assert(HeapTupleIsValid(compressed_tuple));
+	heap_deform_tuple(compressed_tuple,
+					  decompressor->in_desc,
+					  decompressor->compressed_datums,
+					  decompressor->compressed_is_nulls);
+	/* fetch sequence number of current compressed tuple */
+	AttrNumber seq_num =
+		get_attnum(decompressor->in_rel->rd_id, COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME);
+	/* decompress found tuple and delete from compressed chunk */
+	row_decompressor_decompress_row(decompressor, merge_tuplestore);
+	TM_FailureData tmfd;
+	TM_Result result pg_attribute_unused();
+	result = table_tuple_delete(decompressor->in_rel,
+								&compressed_tuple->t_self,
+								GetCurrentCommandId(false),
+								GetTransactionSnapshot(),
+								InvalidSnapshot,
+								true,
+								&tmfd,
+								false);
+	Assert(result == TM_Ok);
+	/* make changes visible */
+	CommandCounterIncrement();
+	return decompressor->compressed_datums[seq_num - 1];
+}
+
+static int
+fetch_compressed_tuples_per_segment(RowDecompressor *decompressor, Chunk *uncompressed_chunk,
+									CompressedSegmentInfo **current_segment, int sequence_num,
+									int nsegmentby_cols, Tuplesortstate *tuplestore)
+{
+	int segments_decompressed = 0;
+	/* build scankeys for segmentby and sequence number columns */
+	ScanKeyData *scankeys = palloc0((nsegmentby_cols + 1) * sizeof(ScanKeyData));
+	int num_scankeys = 0;
+	int seg_idx = 0;
+
+	/* build scan keys to do lookup into compressed chunks */
+	for (int i = 0; i < decompressor->out_desc->natts; i++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(decompressor->out_desc, i);
+		if (attr->attisdropped)
+			continue;
+		FormData_hypertable_compression *fd =
+			ts_hypertable_compression_get_by_pkey(uncompressed_chunk->fd.hypertable_id,
+												  attr->attname.data);
+		if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+		{
+			num_scankeys =
+				create_segment_filter_scankey(decompressor,
+											  attr->attname.data,
+											  BTEqualStrategyNumber,
+											  scankeys,
+											  num_scankeys,
+											  NULL,
+											  current_segment[seg_idx++]->segment_info->val,
+											  false);
+		}
+	}
+	while (true)
+	{
+		sequence_num += SEQUENCE_NUM_GAP;
+		num_scankeys = create_segment_filter_scankey(decompressor,
+													 COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME,
+													 BTEqualStrategyNumber,
+													 scankeys,
+													 num_scankeys,
+													 NULL,
+													 sequence_num,
+													 false);
+		HeapTuple compressed_tuple;
+		TableScanDesc heapScan =
```

Comment:
I was wondering if it might be possible/more beneficial to do an index scan here on the compressed chunk instead of a heapscan?

Reply:
On a compressed chunk, we have an index defined on segmentby column and sequence number. Thus as you say i can use the index to scan the chunk.
Made necessary changes too.